core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -4293,10 +4291,22 @@ public void doExit( StaplerRequest req, StaplerResponse rsp ) throws IOException
             }
         }
 
-        cleanUp();
-        System.exit(0);
-    }
+        new Thread("exit thread") {
+            @Override
+            public void run() {
+                try {
+                    ACL.impersonate(ACL.SYSTEM);
+                    LOGGER.info(String.format("Shutting down VM as requested by %s from %s",
+                            getAuthentication().getName(), req != null ? req.getRemoteAddr() : "???"));
 
+                    cleanUp();
+                    System.exit(0);
+                } catch (Exception e) {
+                    LOGGER.log(Level.WARNING, "Failed to shut down Jenkins", e);
```

Comment:
Probably should be an error, it looks like it's more than just a warning. No big deal though.